### PR TITLE
Expand environment variables everywhere

### PR DESCRIPTION
**Pull request by rgrinberg**
Now they should be allowed in all stanzas and picked up anywhere where we expand
variables.

This is where I'm noticing that this `~scope` argument is really redundant most of the time. It's just used to expand `%{project_root}` and I think the directory should be enough to determine this already.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Okay, this PR is now ready. I will add some tests + change log entry.

**Comment by rgrinberg** [comment]
There's one left issue remaining: `set-env` is not properly interpreted when actually executing actions themselves. The current code simply fetches the environment from the context rather than fetching it from the most relevant env.

To fix this, I think we just need to pass a `~dir` to the `add_rule` family of functions. Then, this argument could be used to fetch the currently environment for the directory and passed along. This is a fairly tedious change, so if anyone has objections, please don't wait :)

**Comment by rgrinberg** [comment]
Okay, the last commit fixes everything. I have a concern with my implementation, however. The new code will add a dependency on all the env vars in the passed `Env.t` as a dependency. This will cause a recompilation on any env var change, which seems to conservative.

I propose that we add a dependency only the env vars that are mentioned in the dune file itself. One simple way to implement that is to add a flag to every key in `Env.t` to mark if it came externally or from a dune file. We'll only add a dependency on the environment variables that came from a dune file.

@diml what do you think?

**Comment by rgrinberg** [comment]
The error in CI is quite bizarre:

```
-  var visible from dune: -principal         foo alias default
-  DUNE_FOO: -principal
+      ocamlopt foo.exe (exit 2)
+  (cd _build/default && /home/travis/ocaml/bin/ocamlopt.opt -principal -g -o foo.exe .foo.eobjs/foo.cmx)
+  gcc: fatal error: -fuse-linker-plugin, but liblto_plugin$ext_dll not found
+  compilation terminated.
+  File "caml_startup", line 1:
+  Error: Error during linking
+  [1]
```

Is the principal flag causing this? I can just fix the test to avoid this by using a different flag, but this seems concerning enough to investigate.

**Comment by diml** [comment]
Depending on the whole environment seems a bit too conservative indeed. I suppose it would be fine if we cleaned up the environment. Recording the source of variables seems fine.

**Comment by diml** [comment]
Regarding the `dir` argument, we should pass it to `Build_interpret.Rule.make` instead of using the directory of the first target.

**Comment by diml** [comment]
No idea what's going on with the CI... I restarted the job.

**Comment by rgrinberg** [comment]
> Depending on the whole environment seems a bit too conservative indeed. I suppose it would be fine if we cleaned up the environment. Recording the source of variables seems fine.

Actually, I think we might have already introduced this conservative dependencies even without this PR. I see that we are currently depending on `Context.env`, but that already includes `Env.initial`.

**Comment by rgrinberg** [comment]
I'll fix the conservative dependencies in a separate PR (with a test) as this one doesn't actually change this behavior.

> Regarding the dir argument, we should pass it to Build_interpret.Rule.make instead of using the directory of the first target.

I gave this a try but it doesn't really make anything cleaner. If anything, it now makes it easier to get the directory wrong as it isn't inferred from the targets. Why do you think this is better?

**Comment by diml** [comment]
It's just that conceptually, I don't understand what the new `dir` argument means if it is not the directory of the targets. How would you describe it?

**Comment by rgrinberg** [comment]
I would it describe it as: The directory in which the rule was defined in. Which determines which environment it should inherit. Note that it's not necessarily the same as the as the directory of the targets. For example, the install rules are also pretty much defined in the context root dir, but the targets they create are in `install/..`.

**Comment by diml** [comment]
Ok, could you add this description in the documentation of `Super_context.add_rule`?

**Comment by diml** [comment]
Hmm, but rules in `install/..` don't need an environment as they are just symlink rules. What are the cases where we need an environment and this `dir` argument doesn't match the targets directory?

**Comment by rgrinberg** [comment]
> Hmm, but rules in install/.. don't need an environment as they are just symlink rules. What are the cases where we need an environment and this dir argument doesn't match the targets directory?

When you're building object files, you want the environment to come from `lib/foo`, but you're generating targets in `lib/foo/.foo.objs`. We could actually use the latter dir as the environment dir, but it just creates a little inefficiency as we'll have to initialize a lot more env nodes in our table. But otherwise, we could certainly use an explicit `~dir` argument for the target dir.

**Comment by diml** [comment]
The only thing that bothers me is that that we need to fill in and maintain one extra bit of information for every single rule. This is one more possible cause for things to go wrong. To some extent, it feels like we want to pass around a directory context that would in particular contain the environment of the directory.

**Comment by rgrinberg** [comment]
I have a commit where I implement this `dir` field on the rule btw. So if you'd like to see how things look like with it, I can simply push it. It's not a big change.

**Comment by diml** [comment]
Yes please

**Comment by rgrinberg** [comment]
Okay, I pushed it. It's the second last commit. Actually, it still needs a bit of work before it's functional. As there's no valid scope defined in the install dir.

**Comment by diml** [comment]
Another way to deal with this would be to do the following change: instead of passing around a `Super_context.t`, we could pass a context that would represent the rule generation context, i.e. it would be composed:

- a super_context
- an optional `Env_node`
- a scope

That would actually help quite a bit for another refactoring that will come when we start using the new computation model. 

However, we shouldn't do this in this PR as it'd become too big. So we can simply continue without adding the `~dir` argument to `Rule.make` for now

**Comment by rgrinberg** [comment]
> Another way to deal with this would be to do the following change: instead of passing around a Super_context.t, we could pass a context that would represent the rule generation context, i.e. it would be composed:

Okay makes sense. I will that do that refactoring next this sounds a lot better than we have now. Why do we need a scope plus an environment? I thought the scope is part of the environment. Also, shouldn't the scope be optional as well? It's only necessary for generating some rules.

Btw, have a look at the last commit which reapplies the fix in #1508 in this PR. The directory is now being passed via the expander when we expand/evaluate anything.

**Comment by rgrinberg** [comment]
Oddly enough, the environment doesn't seem to differ between this PR and master. Continuing to investigate.

**Comment by rgrinberg** [comment]
Okay, everything is fixed. I was simply not correctly inheriting the parent environment. I added a test case just for this. This PR is ready now.

**Comment by rgrinberg** [comment]
By the way, I've just discovered that we are in fact relying on setting the environment to dynamic values. For example:

```
 (action
  (setenv
   NODE
   %{bin:node}
   (chdir
    test-cases/js_of_ocaml
    (progn (run %{exe:cram.exe} -test run.t) (diff? run.t run.t.corrected))))))
```

the above will not work when `%{bin:node}` isn't found, because I've restricted the arguments of `setenv` to be expandable. In this case, the restriction isn't helpful as we'd get a failure anyway. The restriction is useful in other cases however:

```
(setenv FOO %{read:foo} (echo %{env:FOO=bar}))
```

Here, our expander isn't clever enough to figure out that we cannot expand `FOO` until we've read `foo`.

**Comment by diml** [comment]
> Okay makes sense. I will that do that refactoring next this sounds a lot better than we have now. Why do we need a scope plus an environment? I thought the scope is part of the environment. Also, shouldn't the scope be optional as well? It's only necessary for generating some rules.

TBH I haven't thought in details about what will go in there, I guess will see what we need while doing this refactoring.

**Comment by diml** [comment]
BTW, we should be able to support dynamic values in `setenv`, it's just a bit more work as we'd need to an environment map where the values might be dynamic.

File: src/super_context.ml [hunk 1]
```diff
@@ -261,6 +177,34 @@ end = struct
         Exn.code_error "Super_context.Env.get called on invalid directory"
           [ "dir", Path.to_sexp dir ]
 
+  let external_ t ~dir =
+    let rec loop t node =
```
Comment (diml) [inline src/super_context.ml]: Just a note: if we wrote this function this way instead:

```ocaml
let external_ =
  let rec loop t node =
    ...
  in
  fun t ~dir ->
     ...
```

Then we are sure to not accidentally use the wrong `~dir`  inside `loop`.

**Comment by rgrinberg** [comment]
> BTW, we should be able to support dynamic values in setenv, it's just a bit more work as we'd need an environment map where the values might be dynamic.

Is that enough? The problem is that the value of `%{env:FOO}` is different based on where you are in the template. I'm not sure what you mean by dynamic values, but to me it seems like the expansion context needs to have a map of previously expanded values.

**Comment by diml** [comment]
The idea is to maintain a `(unit, string) Build.t String.Map.t` rather than a `string String.Map.t`. Then you can represent the environment needed to interpret `X` in `(setenv var %{read:foo} X)`.

**Comment by diml** [comment]
Or maybe just `static_or_dynamic String.Map.t` where `static_or_dynamic = Static of string | Dynamic` during partial expansion, and then `string String.Map.t` during the final expansion.

**Comment by rgrinberg** [comment]
I've tried to refactor things to work in a clean way but it's too radical of a change and I'm not fully convinced of the results. Instead, I've added the last commit which is a hack but seems to implement the desired semantics. We should revisit this post 1.5.

**Review by ghost — APPROVED** [review]

**Comment by rgrinberg** [comment]
By the way, note that we are not allowing dynamic variable names still. So for example, `(setenv %{read:foo} bar ..)` is still disallowed. We could actually support this, but let's wait for a use case before supporting weird stuff.

**Comment by diml** [comment]
Agreed. Only accepting static variable names seems like a good thing to do.

**Pull request merged**